httpserver/bin/basic_writer_server.dart

```diff
@@ -9,36 +9,34 @@
 import 'dart:io';
 import 'dart:convert';
 
-void main() {
+main() async {
+  var server =
+      await HttpServer.bind(InternetAddress.LOOPBACK_IP_V4, 4049);
+  await for (var req in server) {
+    ContentType contentType = req.headers.contentType;
+    BytesBuilder builder = new BytesBuilder();
 
-  HttpServer.bind(InternetAddress.LOOPBACK_IP_V4, 4049).then((server) {
-    server.listen((req) {
-
-      ContentType contentType = req.headers.contentType;
-      BytesBuilder builder = new BytesBuilder();
-
-      if (req.method == 'POST' &&
-          contentType != null &&
-          contentType.mimeType == 'application/json') {
-        req.listen((buffer) {
+    if (req.method == 'POST' &&
+        contentType != null &&
+        contentType.mimeType == 'application/json') {
+      try {
+        await for (var buffer in req) {
           builder.add(buffer);
-        }, onDone: () {
-          // Write to a file, get the file name from the URI.
-          String jsonString = UTF8.decode(builder.takeBytes());
-          String filename = req.uri.pathSegments.last;
-          new File(filename).writeAsString(jsonString,
-              mode: FileMode.WRITE).then((_) {
-            Map jsonData = JSON.decode(jsonString);
-            req.response.statusCode = HttpStatus.OK;
-            req.response.write('Wrote data for ${jsonData['name']}.');
-            req.response.close();
-          });
-        });
-      } else {
-        req.response.statusCode = HttpStatus.METHOD_NOT_ALLOWED;
-        req.response.write("Unsupported request: ${req.method}.");
-        req.response.close();
+        }
+      } finally {
+        // Write to a file, get the file name from the URI.
+        String jsonString = UTF8.decode(builder.takeBytes());
+        String filename = req.uri.pathSegments.last;
+        await new File(filename).writeAsString(jsonString,
+            mode: FileMode.WRITE);
+        Map jsonData = JSON.decode(jsonString);
+        req.response.statusCode = HttpStatus.OK;
+        req.response.write('Wrote data for ${jsonData['name']}.');
       }
-    });
-  });
+    } else {
+      req.response.statusCode = HttpStatus.METHOD_NOT_ALLOWED;
+      req.response.write("Unsupported request: ${req.method}.");
+    }
+    req.response.close();
```

Comment:
I think I would have left the close next to statusCode and write. It seems like three instructions you always execute together.